server/src/services/tooljet_db_bulk_upload.service.ts

```diff
@@ -0,0 +1,172 @@
+      .transform((row) => this.validateAndParseColumnDataType(internalTableColumnSchema, row, rowsProcessed))
+      .on('data', (row) => {
+        rowsProcessed++;
+        if (row.id) {
+          if (idstoUpdate.has(row.id))
+            throw new BadRequestException(`Duplicate 'id' value found on row [${rowsProcessed + 1}]: ${row.id}`);
+
+          idstoUpdate.add(row.id);
+          rowsToUpdate.push(row);
+        } else {
+          rowsToInsert.push(row);
+        }
+      })
+      .on('error', (error) => {
+        passThrough.emit('error', new BadRequestException(error.message));
+      })
+      .on('end', () => {
+        passThrough.emit('end');
+      });
+
+    await pipeline(passThrough, csvStream);
+
+    await this.tooljetDbManager.transaction(async (tooljetDbManager) => {
+      await this.bulkInsertRows(tooljetDbManager, rowsToInsert, internalTableId);
+      await this.bulkUpdateRows(tooljetDbManager, rowsToUpdate, internalTableId);
+    });
+
+    return { processedRows: rowsProcessed, rowsInserted: rowsToInsert.length, rowsUpdated: rowsToUpdate.length };
+  }
+
+  async bulkUpdateRows(tooljetDbManager: EntityManager, rowsToUpdate: unknown[], internalTableId: string) {
+    if (isEmpty(rowsToUpdate)) return;
+
+    const updateQueries = rowsToUpdate.map((row) => {
+      const columnNames = Object.keys(rowsToUpdate[0]);
+      const setClauses = columnNames
+        .map((column) => {
+          return `${column} = $${columnNames.indexOf(column) + 1}`;
+        })
+        .join(', ');
+
+      return {
+        text: `UPDATE "${internalTableId}" SET ${setClauses} WHERE id = $${columnNames.indexOf('id') + 1}`,
+        values: columnNames.map((column) => row[column]),
+      };
+    });
+
+    for (const updateQuery of updateQueries) {
+      await tooljetDbManager.query(updateQuery.text, updateQuery.values);
+    }
+  }
+
+  async bulkInsertRows(tooljetDbManager: EntityManager, rowsToInsert: unknown[], internalTableId: string) {
+    if (isEmpty(rowsToInsert)) return;
+
+    const rowsWithoutIdColumn = rowsToInsert.map((row: { id: string }) => {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const { id, ...rest } = row;
+      return rest;
+    });
+
+    const columns = Object.keys(rowsWithoutIdColumn[0]);
+    const rowValues = rowsWithoutIdColumn.map((r) => Object.values(r));
+
+    await tooljetDbManager.createQueryBuilder().insert().into(internalTableId, columns).values(rowValues).execute();
+  }
+
+  async validateHeadersAsColumnSubset(internalTableColumnSchema: TableColumnSchema[], headers: string[]) {
+    const internalTableColumns = new Set<string>(internalTableColumnSchema.map((c) => c.column_name));
+    const columnsInCsv = new Set<string>(headers);
+    const isSubset = (subset: Set<string>, superset: Set<string>) => [...subset].every((item) => superset.has(item));
+
+    if (!isSubset(new Set<string>(headers), internalTableColumns)) {
```

Comment:
Instead of `new Set<string>(headers)` we can use `columnsInCsv` a set has been already created